### ISSUE TITLE
微调数据 loss不下降

### ISSUE THREAD

**Issue by ibmxiang** [THUDM/ChatGLM3#238]
### System Info / 系統信息

采取的训练样本数据为：[
  {
    "prompt": "<prompt text>",
    "response": "<response text>"
  }
  // ...
]
这是我的finetune-pt.sh的训练代码。。请问这是哪里出现了问题喃。
PRE_SEQ_LEN=128
LR=1e-3
MAX_SOURCE_LEN=1024
MAX_TARGET_LEN=128
DEV_BATCH_SIZE=1
GRAD_ACCUMULARION_STEPS=1
MAX_STEP=50
SAVE_INTERVAL=500
BASE_MODEL_PATH=/root/models/chatglm3-6b
DATASET_PATH=datas/self_congni.jsonl
OUTPUT_DIR=/root/fine-models/glm3_owner_fine

mkdir -p $OUTPUT_DIR

torchrun --nproc_per_node=2 finetune.py \
    --train_format input-output \
    --train_file $DATASET_PATH \
    --preprocessing_num_workers 1 \
    --model_name_or_path $BASE_MODEL_PATH \
    --output_dir $OUTPUT_DIR \
    --max_source_length $MAX_SOURCE_LEN \
    --max_target_length $MAX_TARGET_LEN \
    --per_device_train_batch_size $DEV_BATCH_SIZE \
    --gradient_accumulation_steps $GRAD_ACCUMULARION_STEPS \
    --max_steps $MAX_STEP \
    --logging_steps 1 \
    --save_steps $SAVE_INTERVAL \
    --learning_rate $LR \
    --pre_seq_len $PRE_SEQ_LEN 2>&1 | tee ${OUTPUT_DIR}/train.log

这是训练的过程：<<<<<<<<<<<< Sanity Check
/usr/local/python3/lib/python3.10/site-packages/transformers/optimization.py:411: FutureWarning: This implementation of AdamW is deprecated and will be removed in a future version. Use the PyTorch implementation torch.optim.AdamW instead, or set `no_deprecation_warning=True` to disable this warning
  warnings.warn(
[INFO|trainer.py:565] 2023-11-09 14:12:02,305 >> max_steps is given, it will override any value given in num_train_epochs
/usr/local/python3/lib/python3.10/site-packages/transformers/optimization.py:411: FutureWarning: This implementation of AdamW is deprecated and will be removed in a future version. Use the PyTorch implementation torch.optim.AdamW instead, or set `no_deprecation_warning=True` to disable this warning
  warnings.warn(
[INFO|trainer.py:1686] 2023-11-09 14:12:03,239 >> ***** Running training *****
[INFO|trainer.py:1687] 2023-11-09 14:12:03,239 >>   Num examples = 80
[INFO|trainer.py:1688] 2023-11-09 14:12:03,239 >>   Num Epochs = 2
[INFO|trainer.py:1689] 2023-11-09 14:12:03,239 >>   Instantaneous batch size per device = 1
[INFO|trainer.py:1692] 2023-11-09 14:12:03,239 >>   Total train batch size (w. parallel, distributed & accumulation) = 2
[INFO|trainer.py:1693] 2023-11-09 14:12:03,239 >>   Gradient Accumulation steps = 1
[INFO|trainer.py:1694] 2023-11-09 14:12:03,239 >>   Total optimization steps = 50
[INFO|trainer.py:1695] 2023-11-09 14:12:03,240 >>   Number of trainable parameters = 1,835,008
  0%|          | 0/50 [00:00<?, ?it/s]11/09/2023 14:12:03 - WARNING - transformers_modules.chatglm3-6b.modeling_chatglm - `use_cache=True` is incompatible with gradient checkpointing. Setting `use_cache=False`...
11/09/2023 14:12:03 - WARNING - transformers_modules.chatglm3-6b.modeling_chatglm - `use_cache=True` is incompatible with gradient checkpointing. Setting `use_cache=False`...
{'loss': 6.2793, 'learning_rate': 0.00098, 'epoch': 0.03}
{'loss': 4.3691, 'learning_rate': 0.00096, 'epoch': 0.05}
{'loss': 5.2051, 'learning_rate': 0.00094, 'epoch': 0.07}
{'loss': 4.4893, 'learning_rate': 0.00092, 'epoch': 0.1}
{'loss': 4.7539, 'learning_rate': 0.0009000000000000001, 'epoch': 0.12}
{'loss': 4.373, 'learning_rate': 0.00088, 'epoch': 0.15}
{'loss': 5.2168, 'learning_rate': 0.00086, 'epoch': 0.17}
{'loss': 5.2012, 'learning_rate': 0.00084, 'epoch': 0.2}
{'loss': 5.6973, 'learning_rate': 0.00082, 'epoch': 0.23}
{'loss': 4.1719, 'learning_rate': 0.0008, 'epoch': 0.25}
{'loss': 4.6797, 'learning_rate': 0.0007800000000000001, 'epoch': 0.28}
{'loss': 3.8857, 'learning_rate': 0.00076, 'epoch': 0.3}
{'loss': 4.6133, 'learning_rate': 0.00074, 'epoch': 0.33}
{'loss': 5.1543, 'learning_rate': 0.0007199999999999999, 'epoch': 0.35}
{'loss': 4.4199, 'learning_rate': 0.0007, 'epoch': 0.38}
{'loss': 4.1836, 'learning_rate': 0.00068, 'epoch': 0.4}
{'loss': 2.9062, 'learning_rate': 0.00066, 'epoch': 0.42}
{'loss': 4.0898, 'learning_rate': 0.00064, 'epoch': 0.45}
{'loss': 4.2812, 'learning_rate': 0.00062, 'epoch': 0.47}
{'loss': 3.1943, 'learning_rate': 0.0006, 'epoch': 0.5}
{'loss': 3.4668, 'learning_rate': 0.00058, 'epoch': 0.53}
{'loss': 4.209, 'learning_rate': 0.0005600000000000001, 'epoch': 0.55}
{'loss': 3.8887, 'learning_rate': 0.00054, 'epoch': 0.57}
{'loss': 4.1484, 'learning_rate': 0.0005200000000000001, 'epoch': 0.6}
{'loss': 6.2793, 'learning_rate': 0.0005, 'epoch': 0.62}
{'loss': 3.417, 'learning_rate': 0.00048, 'epoch': 0.65}
{'loss': 4.3867, 'learning_rate': 0.00046, 'epoch': 0.68}
{'loss': 5.0684, 'learning_rate': 0.00044, 'epoch': 0.7}
{'loss': 3.9365, 'learning_rate': 0.00042, 'epoch': 0.72}
{'loss': 4.5059, 'learning_rate': 0.0004, 'epoch': 0.75}
{'loss': 3.4688, 'learning_rate': 0.00038, 'epoch': 0.78}
{'loss': 4.7793, 'learning_rate': 0.00035999999999999997, 'epoch': 0.8}
{'loss': 3.7539, 'learning_rate': 0.00034, 'epoch': 0.82}
{'loss': 5.7383, 'learning_rate': 0.00032, 'epoch': 0.85}
{'loss': 3.9961, 'learning_rate': 0.0003, 'epoch': 0.88}
{'loss': 4.4883, 'learning_rate': 0.00028000000000000003, 'epoch': 0.9}
{'loss': 4.6035, 'learning_rate': 0.00026000000000000003, 'epoch': 0.93}
{'loss': 4.3965, 'learning_rate': 0.00024, 'epoch': 0.95}
{'loss': 4.168, 'learning_rate': 0.00022, 'epoch': 0.97}
{'loss': 3.1309, 'learning_rate': 0.0002, 'epoch': 1.0}
{'loss': 4.8633, 'learning_rate': 0.00017999999999999998, 'epoch': 1.02}
{'loss': 3.791, 'learning_rate': 0.00016, 'epoch': 1.05}
{'loss': 3.9297, 'learning_rate': 0.00014000000000000001, 'epoch': 1.07}
{'loss': 5.4082, 'learning_rate': 0.00012, 'epoch': 1.1}
{'loss': 4.1836, 'learning_rate': 0.0001, 'epoch': 1.12}
{'loss': 4.2461, 'learning_rate': 8e-05, 'epoch': 1.15}
{'loss': 4.4766, 'learning_rate': 6e-05, 'epoch': 1.18}
{'loss': 3.2949, 'learning_rate': 4e-05, 'epoch': 1.2}
{'loss': 4.4277, 'learning_rate': 2e-05, 'epoch': 1.23}
{'loss': 3.7422, 'learning_rate': 0.0, 'epoch': 1.25}
100%|██████████| 50/50 [00:52<00:00,  1.05s/it][INFO|trainer.py:1934] 2023-11-09 14:12:55,816 >> 

### Information / 问题信息

- [X] The official example scripts / 官方的示例脚本
- [ ] My own modified scripts / 我自己修改的脚本和任务

### Reproduction / 复现过程

无

### Expected behavior / 期待表现

希望能够成功的微调结果。输入prompt，得到训练集中response结果

**Comment by kamieee** [comment]
我的loss也差不多4-6徘徊

**Comment by sunjiadai** [comment]
我会下降

**Comment by ibmxiang** [comment]
> 我会下降

麻烦给个几行训练示例，跟训练参数喃

**Comment by zRzRzRzRzRzRzR** [comment]
loss不降的解决方案：
1. 降低学习率
2. 根据你的数据量确定一个合适的学习轮次
3. 检查一下数据的结构和内容，确保具有一定关联
训练示例暂时还不能给出，之后有更新的教程会进行推进，敬请期待

**Comment by morettt** [comment]
你看看MAX_SOURCE_LEN=1024
MAX_TARGET_LEN=128  这两个参数。一个是输入MAX_SOURCE_LEN，也就是数据集的prompt的那部分的token,这里你填的是1024 另一个MAX_TARGET_LEN指的是输出，这里就是数据集的response那部分，这里是128, 一般来说这两个最好是平衡。最好去检查一下你的数据集输入输出的token是不是在这个区间内。不然会浪费很多的资源，还会影响模型。我现在才发现，之前一直没有去关注这个参数，浪费了好多的时间。
